### ISSUE TITLE
Memory leak in PerfCounterQueryHandler constructor

### ISSUE THREAD

**Issue by Sebastian-0** [oshi/oshi#1244]
When running my oshi-based task manager for a long time (on Windows) it starts to use more and more memory and eventually fails with OOM. I did some profiling and narrowed the issue down to `PerfCounterQueryHandler`. The problem is that every time you update CPU ticks, disk attributes, etc... a new `PerfCounterQueryHandler` is created and registered in a shutdown hook. The hook will be kept alive until the termination of the JVM with a reference to the object, keeping it from being garbage collected and creating a memory leak. Obviously it takes a long time before you notice it but when I deploy my program it can run for weeks at a time which makes this a big problem.

For your reference I have a small code example that illustrates the issue if you leave it running for 10-20 minutes (the same one from my previous issue):
```
public class MemTest {
	public static void main(String[] args) {
		SystemInfo info = new SystemInfo();
		List<HWDiskStore> disks = info.getHardware().getDiskStores();
		while (true) {
			for (HWDiskStore disk : disks) {
				disk.updateAttributes();
			}
		}
	}
}
```

This causes the following heap usage over time (the dips are me forcing GCs manually):
![image](https://user-images.githubusercontent.com/6851439/83325780-62bf5f80-a26f-11ea-84a5-0486c7d1dbb2.png)

Doing some memory-sampling shows:
![image](https://user-images.githubusercontent.com/6851439/83325912-538ce180-a270-11ea-9102-d7a40941e2a2.png)
where the top usage is caused by `PerfCounterQueryHandler`, its concurrent hash maps and probably their contents (all the strings).

**Comment by dbwiddis** [comment]
Hmm.  The shutdown hook was a relic of when the map was a static object.  So basically the only thing keeping the object alive is something saying "Hey, don't forget to release me when you're done"?

I think the correct fix is just to remove that shutdown hook.

**Comment by dbwiddis** [comment]
No, not quite that simple.  There is some cleanup that is supposed to be done.  The quick fix would be to move that code to a finalize method.  But <insert list of reasons not to use finalize here>.    So what I really need to do is use close, AutoCloseable, and try-with-resources blocks.

Well, there went my weekend. :)

**Comment by Sebastian-0** [comment]
Sorry about your weekend :P 

I think it would be best to remove the shutdown hook and make sure the resources are closed properly like you suggest. If the rest of the code properly handles your resources you shouldn't need the hook as a backup.

I hope it's not hard to fix, I considered creating a PR but I was not sure how you want that part of the code base to work :)

**Comment by dbwiddis** [comment]
It's not too hard. I'm completely gutting the existing PerfDataQuery classes, making stateless, static query methods.  Makes them thread-safe in addition to fixing this leak.

Pretty much done with the rewrite, then to test the rewrite works and remove the old version. :)

I hope you'll be available to test it in an hour or so.

**Comment by dbwiddis** [comment]
OK, this should be fixed in #1245, available in SNAPSHOT.  Will do some more testing and doc updates on this end before merging.

**Comment by Sebastian-0** [comment]
I will give it a go, reporting back soon!

**Comment by Sebastian-0** [comment]
This fixes my issue, there are no longer any memory leaks.

**Comment by dbwiddis** [comment]
Great!  Merging the fix and will cut a 5.1.1 release within 24 hours.
